feat(AIP-133) up resource_id guidance to must

toumorokoshi opened the pull request:
The current wording implies that user-settable resource_ids are somewhat helpful *may*, when in most cases they are very beneficial: they are critical for declarative clients to uniquely identify resources and prevent duplicates, and enable the user to provide semantic names and organization for resources.

Generally the guidance would be better for APIs if they were added if at all possible, and only skipped if there the usage pattern is exceptional (e.g. job resources with significant churn, revisions where a unique identifier is determined by the custom method).

toumorokoshi committed:
feat(AIP-133) up resource_id guidance to should

The current wording implies that user-settable resource_ids
are somewhat helpful *may*, when in most cases they are very
beneficial: they are critical for declarative clients to uniquely
identify resources and prevent duplicates, and enable the user
to provide semantic names and organization for resources.

Generally the guidance would be better for APIs if they were
added if at all possible, and only skipped if there the usage
pattern is exceptional (e.g. job resources with significant churn,
revisions where a unique identifier is determined by the custom
method).

toumorokoshi committed:
Merge branch 'master' into toum/resource-id-to-should

toumorokoshi committed:
addressing feedback

- make user-settable IDs a must.
- add rationale

toumorokoshi committed:
Merge branch 'master' into toum/resource-id-to-should

toumorokoshi committed:
address feedback

- removed must on resource_id in AIP-128

toumorokoshi committed:
more wording tweaks

- clarifying and fixing user-settable ID section.

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -79,6 +79,13 @@ message CreateBookRequest {
+  //
+  // This value should be 4-63 characters, and valid characters
+  // are /[a-z][0-9]-/.
+  string book_id = 3 [(google.api.field_behavior) = REQUIRED];
```
Kind of a nit, but I'd put this as the 2nd field.

toumorokoshi replied:
done! It's generally not because of the fact that is originally optional, but
should be fine now.

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -88,6 +95,7 @@ message CreateBookRequest {
   - The field **must** identify the [resource type][aip-123] of the resource
     being created.
 - The resource field **must** be included and **must** map to the POST body.
+- A `{resource}_id` field **must** be included.
```
I'd insert this above the previous line also.

toumorokoshi replied:
ack

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -194,17 +175,32 @@ NOT_FOUND errors][permission-denied].
 - For ensuring idempotency in `Create` methods, see [AIP-155][].
```
We should remove this line and change AIP 155 to no longer be recommended.

toumorokoshi replied:
tracking issue: https://github.com/aip-dev/google.aip.dev/issues/1103

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -194,17 +175,32 @@ NOT_FOUND errors][permission-denied].
 
+## Rationale
+
+### Requiring user-settable ids
```
I like to say "client-settable", which is more general than "user". It could be generated by a client, such as a controller.

toumorokoshi replied:
user-settable was the original nomenclature so I was sticking to that, but I personally also like the term "client" a bit better. will change.

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -194,17 +175,32 @@ NOT_FOUND errors][permission-denied].
+
+### Requiring user-settable ids
+
+[IaC][] clients use the resource ID as a way to identify a resource for applying
```
And other clients.

This basically replaces the approach recommended by https://google.aip.dev/155 to achieve idempotence for Create.

Using the resource id is simpler for clients and for services. It eliminates the need for a request id separate from the resource id. Resource ids already need to be unique within a project and location.

If the resource id isn't known until, say, a LRO completes, that widens the window during which a failure could cause the client to not know whether it had already issued the Create request and whether it had been successful or not. AIP 155 doesn't specify that the LRO needs to return the request id, either.

If the service needs resource ids to be of a particular form, it can enforce that with validation. We need better ways to convey such value constraints to users and clients, though.

Note that the desire for idempotence makes the practice of allowing the id to be optional fairly dubious. It would be pointless to make the resource id optional and then require an additional request id instead, for example.

toumorokoshi replied:
Those are great points! I'll add them to the PR. 

I think I buy the idea that we can do away with 155 entirely, and for those that want idempotency they need to provide a resource ID.

I think there's still good use cases for having an optional resource ID. e.g. cloud build triggers generating an ID on your behalf. But agreed that there should not be a concept of ensuring idempotency of an anonymous create request.

toumorokoshi replied:
That said I think removal of 155 needs be in a separate PR due to the size of the scope there: I've added #1103 for tracking.

toumorokoshi replied:
WDYT about me adding this to with the PR for AIP-155? I feel like we need to
make to change to remove 155 and replace it in lockstep, or else there's
conflicting guidance.

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -194,17 +175,32 @@ NOT_FOUND errors][permission-denied].
+and can result in re-creating the resource.
+
+Having a user-settable ID also means the client can precalculate the resource
+name and use it in resource references.
```
in other resources

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -194,17 +175,32 @@ NOT_FOUND errors][permission-denied].
+[IaC][] clients use the resource ID as a way to identify a resource for applying
+updates and for conflict resolution. The lack of a user-settable ID means a
+client is unable to find the resource unless they store the identifier locally,
+and can result in re-creating the resource.
```
There's also the issue of the id changing in the case that the resource is deleted and re-created, which could also require changing all resources that refer to it.

toumorokoshi replied:
added!

toumorokoshi committed:
addressing feedback

user-settable -> client-settable.

Adding more rationale, fixing ordering so ID is required first,
then the body of the resource.

toumorokoshi committed:
Merge branch 'master' into toum/resource-id-to-should

toumorokoshi committed:
Merge branch 'master' into toum/resource-id-to-should

bgrant0607 reviewed (approved):
Thanks. LGTM

toumorokoshi committed:
Merge branch 'master' into toum/resource-id-to-should

alin04 commented on aip/general/0133.md:
```diff
@@ -58,10 +58,8 @@ rpc CreateBook(CreateBookRequest) returns (Book) {
   **must** map to the resource field in the request message.
   - All remaining fields **should** map to URI query parameters.
 - There **should** be exactly one `google.api.method_signature` annotation,
```
Why is this **should** and not **must** ?

toumorokoshi replied:
@jskeet may be able to speak to this. I don't see any reason why it shouldn't be a must, but this affects client libraries mainly. 

for other clients, this annotation doesn't mean anything.

toumorokoshi replied:
as a quick note: this was **should** originally, so it's actually just adjusting the terminology based on  the new required value, rather than adjusting the level of guidance.

alin04 replied:
Ack. Can revisit should vs must separately.
